parse/types.go

```diff
@@ -0,0 +1,192 @@
+		StartCol:  end,
+		EndLine:   start,
+		EndCol:    end,
+	}
+}
+
+// MergeNodes merges two Positions into a single Position.
+// It starts at the left Position and ends at the right Position.
+func MergeNodes(left, right *Position) *Position {
+	return &Position{
+		StartLine: left.StartLine,
+		StartCol:  left.StartCol,
+		EndLine:   right.EndLine,
+		EndCol:    right.EndCol,
+	}
+}
+
+// SchemaInfo contains information about a parsed schema
+type SchemaInfo struct {
+	// Blocks maps declared block names to their Positions.
+	// Block names include:
+	// - tables
+	// - extensions
+	// - actions
+	// - procedures
+	// - foreign procedures
+	Blocks map[string]*Block `json:"blocks"`
+}
+
+type Block struct {
+	Position
+	// AbsStart is the absolute start position of the block in the source code.
+	AbsStart int `json:"abs_start"`
+	// AbsEnd is the absolute end position of the block in the source code.
+	AbsEnd int `json:"abs_end"`
+}
+
+// Relation represents a relation in a sql statement.
+// It is meant to represent the shape of a relation, not the data.
+type Relation struct {
+	Name string
+	// Attributes holds the attributes of the relation.
+	Attributes []*Attribute
+}
+
+// ShapesMatch checks if the shapes of two relations match.
+// It only checks types and order, and ignores the names of the attributes,
+// as well as uniques and primary keys.
+func (r *Relation) ShapesMatch(r2 *Relation) bool {
+	return ShapesMatch(r.Attributes, r2.Attributes)
+}
+
+// FindAttribute finds an attribute by name.
+// If it is not found, it returns nil and false
+func (r *Relation) FindAttribute(name string) (*Attribute, bool) {
+	for _, a := range r.Attributes {
+		if a.Name == name {
+			return a, true
+		}
+	}
+	return nil, false
+}
+
+// Copy returns a copy of the relation.
+func (r *Relation) Copy() *Relation {
+	attrs := make([]*Attribute, len(r.Attributes))
+	for i, a := range r.Attributes {
+		attrs[i] = &Attribute{
+			Name: a.Name,
+			Type: a.Type.Copy(),
+		}
+	}
+
+	return &Relation{
+		Name:       r.Name,
+		Attributes: attrs,
+	}
+}
+
+// Attribute represents an attribute in a relation.
+type Attribute struct {
+	Name string
+	Type *types.DataType
+}
+
+// ShapesMatch checks if the shapes of two relations match.
+// It only checks types and order, and ignores the names of the attributes,
+// as well as uniques and primary keys.
+func ShapesMatch(a1, a2 []*Attribute) bool {
+	if len(a1) != len(a2) {
+		return false
+	}
+	for i := range a1 {
+		if !a1[i].Type.Equals(a2[i].Type) {
+			return false
+		}
+	}
+	return true
+}
+
+// Flatten flattens many relations into one unnamed relation.
+// If there are column conflicts / ambiguities, it will return an error,
+// and the name that caused the conflict. It will also discard any
+// primary keys, and will unmark any unique columns.
+func Flatten(rels ...*Relation) (res []*Attribute, col string, err error) {
+	for _, r := range rels {
+		res, col, err = Coalesce(append(res, r.Attributes...)...)
```

Comment:
It looks we'll rerun multiple times of the rel from earlier loop.  We can call `Coalesce` outside of the loop

Reply:
I don't see how we can call `Coalesce` outside of the loop. Will fix the rerun issue

Reply:
Oh never mind I see